src/engine/srv.c

```diff
@@ -843,31 +910,149 @@ dss_xstream_is_busy(void)
 	return cur_msec < (cntr->rc_active_time + 5000);
 }
 
+struct core_assignment {
+	int	*ca_assigned;			/* Assigned core idx array */
+	int	 ca_sub_off[DSS_XS_MAX];	/* Offset of ca_assigned for XS types */
+	int	 ca_sub_nr[DSS_XS_MAX];		/* Number of assigned for XS types */
+};
+
+static void
+ca_assign(struct core_assignment *ca, unsigned int xs_type, unsigned idx)
+{
+	int	sub_off, sub_nr;
+
+	D_ASSERT(xs_type < DSS_XS_MAX);
+
+	/* IOFW and OFFLOAD share the same sub-array */
+	if (xs_type == DSS_XS_OFFLOAD)
+		xs_type = DSS_XS_IOFW;
+	sub_off = ca->ca_sub_off[xs_type];
+	sub_nr = ca->ca_sub_nr[xs_type];
+
+	D_ASSERT(ca->ca_assigned[sub_off + sub_nr] == -1);
+	ca->ca_assigned[sub_off + sub_nr] = idx;
+	ca->ca_sub_nr[xs_type] += 1;
+}
+
+static void
+ca_free(struct core_assignment *ca)
+{
+	D_FREE_PTR(ca->ca_assigned);
+	D_FREE_PTR(ca);
+}
+
+static struct core_assignment *
+ca_alloc(unsigned int xs_nr)
+{
+	struct core_assignment	*ca;
+	int			 i, off = 0;
+
+	D_ALLOC_PTR(ca);
+	if (ca == NULL)
+		return NULL;
+
+	D_ALLOC_ARRAY(ca->ca_assigned, xs_nr);
+	if (ca->ca_assigned == NULL) {
+		D_FREE(ca);
+		return NULL;
+	}
+
+	for (i = 0; i < xs_nr; i++)
+		ca->ca_assigned[i] = -1;
+
+	ca->ca_sub_off[DSS_XS_SYS]	= off;
+	off += 1;
+
+	ca->ca_sub_off[DSS_XS_SWIM]	= off;
+	off += 1;
+
+	ca->ca_sub_off[DSS_XS_DRPC]	= off;
+	off += DRPC_XS_NR;
+
+	ca->ca_sub_off[DSS_XS_VOS]	= off;
+	off += dss_tgt_nr;
+
+	ca->ca_sub_off[DSS_XS_IOFW]	= off;
+	ca->ca_sub_off[DSS_XS_OFFLOAD]	= off;
+	off += dss_tgt_offload_xs_nr;
+
+	ca->ca_sub_off[DSS_XS_SEC]	= off;
+	off += dss_sec_xs_nr;
+	D_ASSERT(off == xs_nr);
+
+	return ca;
+}
+
+/* Reuse a core form the cores assigned to 'assigned_type' xstream */
+static unsigned
+ca_reuse_core(struct core_assignment *ca, unsigned int xs_id, unsigned int assigned_type)
+{
+	int sub_off, sub_nr, idx;
+
+	sub_off = ca->ca_sub_off[assigned_type];
+	sub_nr = ca->ca_sub_nr[assigned_type];
+
+	D_ASSERT(sub_nr > 0);
+	if (sub_nr > 1)
+		sub_off += (xs_id % sub_nr);
+
+	idx = ca->ca_assigned[sub_off];
+	D_ASSERT(idx != -1);
+
+	return idx;
+}
+
+/* There are not enough cores for DRPC and SEC xstreams */
+static inline bool
+insufficient_cores()
+{
+	if (numa_obj)
+		return (dss_num_cores_numa_node < DSS_XS_NR_TOTAL);
+	else
+		return (dss_core_nr < DSS_XS_NR_TOTAL);
```

Comment:
I think 'dss_core_offset' is valid for legacy mode only, and looks it affects only the starting idx, all cores are available since we use the 'core_offset % core_nr' to assign cores in legacy mode, right?

Reply:
ah, ok.